### ISSUE TITLE
Allow overloading on type name

### ISSUE THREAD

**Issue by narimiran** [nim-lang/rfcs#3]
<a href="https://github.com/flaviut"><img src="https://avatars3.githubusercontent.com/u/5213469?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [flaviut](https://github.com/flaviut)**
_Monday Dec 01, 2014 at 02:40 GMT_
_Originally opened as https://github.com/nim-lang/Nim/issues/1693_

----

It's impossible to overload on a type name, ex.

``` nimrod
type
  Foo = object
    a: int

proc Foo(): Foo =  # fails to compile
  return Foo(a : 213)
proc Foo(val: int): Foo =  # also fails to compile
  return Foo(a : val)
```

Ideally, this would be possible and possibly even unified with the type construction syntax.

**Comment by narimiran** [comment]
<a href="https://github.com/Araq"><img src="https://avatars1.githubusercontent.com/u/287816?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [Araq](https://github.com/Araq)**
_Monday Dec 01, 2014 at 22:09 GMT_

----

IMHO we shouldn't do this, but I'm leaving it open so people can discuss.

**Comment by narimiran** [comment]
<a href="https://github.com/kirbyfan64"><img src="https://avatars1.githubusercontent.com/u/1690697?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [kirbyfan64](https://github.com/kirbyfan64)**
_Monday Dec 01, 2014 at 22:17 GMT_

----

If this _were_ to be added, I'd prefer it to be like destructors:

``` nim
proc create(): Foo =
    return Foo(a: 213)
...
```

For consistency reasons.

**Comment by narimiran** [comment]
<a href="https://github.com/PavelVozenilek"><img src="https://avatars3.githubusercontent.com/u/5585037?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [PavelVozenilek](https://github.com/PavelVozenilek)**
_Tuesday Dec 02, 2014 at 02:25 GMT_

----

~~One name has several different meanings. What could go possibly wrong?~~~

Edit: changed opinion. Added a comment down bellow.

**Comment by narimiran** [comment]
<a href="https://github.com/flaviut"><img src="https://avatars3.githubusercontent.com/u/5213469?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [flaviut](https://github.com/flaviut)**
_Tuesday Dec 02, 2014 at 03:08 GMT_

----

@kirbyfan64 Nim has a long tradition of using overloading, so it would be consistent. I wouldn't mind the `create()` sort of syntax though.

@PavelVozenilek Of course the programmer should exercise discretion when using it, but Nim already provides a thousand different ways to shoot yourself in the foot, what's one more? :wink:

**Comment by narimiran** [comment]
<a href="https://github.com/superfunc"><img src="https://avatars0.githubusercontent.com/u/1583002?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [superfunc](https://github.com/superfunc)**
_Tuesday Dec 02, 2014 at 04:19 GMT_

----

@kirbyfan64 isn't this just a regular function as we already have?

**Comment by narimiran** [comment]
<a href="https://github.com/kirbyfan64"><img src="https://avatars1.githubusercontent.com/u/1690697?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [kirbyfan64](https://github.com/kirbyfan64)**
_Tuesday Dec 02, 2014 at 19:47 GMT_

----

@superfunc No; It'd work like [destructors](http://nim-lang.org/manual.html#destructors) do. With destructors, Nim automatically invokes the `destroy` function with the object as its argument. With the syntax I showed, it's the same basic idea; Nim invokes the `create` function implicitly; you never see it happen.

Basically, `create` is to `destroy` as constructors are to destructors.

**Comment by narimiran** [comment]
<a href="https://github.com/superfunc"><img src="https://avatars0.githubusercontent.com/u/1583002?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [superfunc](https://github.com/superfunc)**
_Tuesday Dec 02, 2014 at 21:53 GMT_

----

That makes sense, my bad

**Comment by narimiran** [comment]
<a href="https://github.com/PavelVozenilek"><img src="https://avatars3.githubusercontent.com/u/5585037?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [PavelVozenilek](https://github.com/PavelVozenilek)**
_Monday Dec 15, 2014 at 03:31 GMT_

----

I think there's use case for this. 

Say I define a type and want to limited functionality on it, via templates

```
# --- a_lib.nim ---
type
  MyFile* = object
     name : string
     handle : int

template foo*(ff : expr, s : string, body : stmt) {.immediate.} =
  block:
    var ff : MyFile
    ff.name = s
    ff.handle = 0
    echo ("name = ", ff.name)
    body

# --- main.nim
import a_lib

foo(f, "name"): 
  echo "Do something with 'f' of being type MyFile"
```

Here I need to use helper name `foo`. If I would be allowed to use ``MyFile` it would mean one superficial name down:

```
# --- main.nim
import a_lib

MyFile(f, "name"): 
  echo "Do something with 'f' of being type MyFile"
```

**Comment by narimiran** [comment]
<a href="https://github.com/flaviut"><img src="https://avatars3.githubusercontent.com/u/5213469?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [flaviut](https://github.com/flaviut)**
_Tuesday Jan 27, 2015 at 01:13 GMT_

----

Related: http://forum.nim-lang.org/t/703

**Comment by narimiran** [comment]
<a href="https://github.com/Henry"><img src="https://avatars1.githubusercontent.com/u/9874?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [Henry](https://github.com/Henry)**
_Sunday Aug 16, 2015 at 09:10 GMT_

----

I would like to see this feature added as it provides consistency between the construction of built-in types and user-defined types.  Also it improves readability of the code when objects are constructed as arguments for function calls.

**Comment by narimiran** [comment]
<a href="https://github.com/Araq"><img src="https://avatars1.githubusercontent.com/u/287816?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [Araq](https://github.com/Araq)**
_Monday Aug 17, 2015 at 12:36 GMT_

----

> I would like to see this feature added as it provides consistency between the construction of built-in types and user-defined types.

There is no inconsistency right now, the compiler treats every type the same.

**Comment by narimiran** [comment]
<a href="https://github.com/Henry"><img src="https://avatars1.githubusercontent.com/u/9874?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [Henry](https://github.com/Henry)**
_Monday Aug 17, 2015 at 12:55 GMT_

----

Sure, but if I need to create a special constructor for a user-defined type it has to be named differently whereas constructors for built-in types are named the same as the type.  If you were to add another constructor for a built-in type would you also name it the same as the type or differently?

**Comment by narimiran** [comment]
<a href="https://github.com/Araq"><img src="https://avatars1.githubusercontent.com/u/287816?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [Araq](https://github.com/Araq)**
_Monday Aug 17, 2015 at 13:29 GMT_

----

I think you're misunderstanding what `int(x)` means: It is a type **conversion**, not a type **construction**, it always is, there is no special logic in the compiler that treats `int` differently from object `Foo`.

> If you were to add another constructor for a built-in type would you also name it the same as the type or differently?

One cannot name it the same.

**Comment by narimiran** [comment]
<a href="https://github.com/Henry"><img src="https://avatars1.githubusercontent.com/u/9874?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [Henry](https://github.com/Henry)**
_Monday Aug 17, 2015 at 14:19 GMT_

----

I am referring to constructors of the form:

```
var student = Student(name: "Anton", age: 5, id: 3)
```

I would like to be able to create other constructors for object types and it would be convenient and, at least to me, consistent  to also name them the same.

> One cannot name it the same.

Right, currently not, but if you could would you?  If not what convention would you use and why should the other constructors be named differently given that overloading is used for other functions?

If you/others are interested in having support for this overloading do you have some idea of how much effort it would take to implement it?

**Comment by narimiran** [comment]
<a href="https://github.com/Varriount"><img src="https://avatars0.githubusercontent.com/u/524596?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [Varriount](https://github.com/Varriount)**
_Monday Aug 17, 2015 at 14:53 GMT_

----

@Henry Those aren't constructors, at least, not in the usual sense. It's merely syntactic sugar for simple object initialization. I don't see how the syntax differs between user types and built-in types.

**Comment by narimiran** [comment]
<a href="https://github.com/Henry"><img src="https://avatars1.githubusercontent.com/u/9874?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [Henry](https://github.com/Henry)**
_Monday Aug 17, 2015 at 15:22 GMT_

----

OK, sure, they do not create a location or allocate storage, they are basically initializers.  So how do a create a set of object initializers which have the same name as the object type or provide the syntactic sugar so that they appear so?  Basically how do I write types so that they look and behave like built-in types but with additional initializers?  This is the consistency I was referring to.

**Comment by narimiran** [comment]
<a href="https://github.com/Varriount"><img src="https://avatars0.githubusercontent.com/u/524596?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [Varriount](https://github.com/Varriount)**
_Monday Aug 17, 2015 at 15:28 GMT_

----

I still don't see what you mean by "like built-in types". Built-in types like integers are assigned to directly, while sequences and strings are created using regular procedures, like `newSeq` and `newString`. Perhaps you could include a comparative example?

**Comment by narimiran** [comment]
<a href="https://github.com/Henry"><img src="https://avatars1.githubusercontent.com/u/9874?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [Henry](https://github.com/Henry)**
_Monday Aug 17, 2015 at 15:41 GMT_

----

Take the example 

```
var student = Student(name: "Anton", age: 5, id: 3)
```

This is a user-defined object type initialized using the default/builtin initializer.  Now imagine Student could be created/initialized in other ways than simply providing all the elements as arguments:

```
var student2 = Student("Anton", database)
```

OK this is a trivial example but in the applications I would like to translate from C++ to Nim there are lots of complex classes which can be constructed/initialized is many ways and in practice it is convenient if these functions name the same as the class (standard C++ constructor naming) particularly when constructing a temporary instance as an argument to a function e.g.:

```
var success = process(Student("Anton", database))
```

**Comment by narimiran** [comment]
<a href="https://github.com/dom96"><img src="https://avatars0.githubusercontent.com/u/246651?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [dom96](https://github.com/dom96)**
_Monday Aug 17, 2015 at 16:00 GMT_

----

The convention for constructors is to define a new proc named `initType` if it creates a non-ref type and `newType` if it creates a ref type.

**Comment by narimiran** [comment]
<a href="https://github.com/Henry"><img src="https://avatars1.githubusercontent.com/u/9874?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [Henry](https://github.com/Henry)**
_Monday Aug 17, 2015 at 16:10 GMT_

----

I understand, in which case shouldn't we use:

```
var student = initStudent(name: "Anton", age: 5, id: 3)
```

Why is the name of the type used for built-in initializers and initType for user-defined initializers?  Wouldn't it be better to overload the type name?  Do you prefer

```
var success = process(initStudent("Anton", database))
```

or

```
var success = process(Student("Anton", database))
```

**Comment by narimiran** [comment]
<a href="https://github.com/Bulat-Ziganshin"><img src="https://avatars1.githubusercontent.com/u/1798016?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [Bulat-Ziganshin](https://github.com/Bulat-Ziganshin)**
_Sunday Jul 15, 2018 at 16:09 GMT_

----

Can we consider type definition as automatically defining magic function(s) with the same name constructing object from field values? :

```Nim
type Foo = tuple a,b: int

# defines the magic function
proc Foo(a,b: int): Foo

type Bar = object case kind: bool of
    true: n: int
    false: b: bool

# defines the magic functions
proc Bar( kind:static[true], n: int): Bar
proc Bar( kind:static[false], b: bool): Bar
```

and user can provide other overloads of Foo/Bar and even overwrite these default definitions with his own ones (or just discard them).

Also, my arguments for using Foo/Bar as object constructors: https://github.com/nim-lang/Nim/issues/7474#issuecomment-405100389

**Comment by narimiran** [comment]
<a href="https://github.com/dom96"><img src="https://avatars0.githubusercontent.com/u/246651?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [dom96](https://github.com/dom96)**
_Sunday Jul 15, 2018 at 21:19 GMT_

----

That's already the case...

```nim
Bar(kind: true, n: 42)
```

https://nim-lang.org/docs/manual.html#types-object-construction

**Comment by narimiran** [comment]
<a href="https://github.com/Bulat-Ziganshin"><img src="https://avatars1.githubusercontent.com/u/1798016?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [Bulat-Ziganshin](https://github.com/Bulat-Ziganshin)**
_Sunday Jul 15, 2018 at 22:20 GMT_

----

Well, my point - can we just add these auto-generated definitions to the procedures list and then allow user to provide other procedures with the same name but different parameters? What is the catch - is it hard to implement or it doesn't look useful?

And yeah, currently object constructor require field names. For Algebraic datatypes similar to Haskell it will be great to allow using these constructors without field names.

Similarly, it will be great to allow to use field names in tuple constructor. I.e. in both cases generate a procedure with the same name as type and parameter names equal to the field names, and allow to overload this procedure.
